server/commands/write_authzmodel.go

```diff
@@ -163,3 +180,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
 
 	return nil
 }
+
+func areRelationalTypesValid(tds []*openfgapb.TypeDefinition) error {
+	typeToRelations := map[string]map[string]struct{}{}
+	for _, td := range tds {
+		typeName := td.GetType()
+		typeToRelations[typeName] = map[string]struct{}{}
+		for relationName := range td.GetRelations() {
+			typeToRelations[typeName][relationName] = struct{}{}
+		}
+	}
+
+	// Here we are checking that every type on a relation is valid. This means:
+	// 1. If it is a type (e.g. user) then this type is a type in the model
+	// 2. If it is a type#relation then this type with this relation is in the model
+	for _, td := range tds {
+		for _relation := range td.GetRelations() {
+			metadata, ok := td.GetMetadata().GetRelations()[_relation]
+			if ok {
+				for _, userType := range metadata.GetDirectlyRelatedUserTypes() {
+					objectType := userType.GetType()
+					relation := userType.GetRelation()
+
+					if _, ok := typeToRelations[objectType]; !ok {
+						return serverErrors.InvalidRelationType(td.GetType(), _relation, objectType, relation)
+					}
+
+					if relation != "" {
+						if _, ok := typeToRelations[objectType][relation]; !ok {
+							return serverErrors.InvalidRelationType(td.GetType(), _relation, objectType, relation)
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// Finally we check that a relation is assignable if and only if it has a non-zero AssignableRelations array
+	for _, td := range tds {
+		for relation, rewrite := range td.GetRelations() {
+			metadata, ok := td.GetMetadata().GetRelations()[relation]
+			if isAssignable(rewrite) {
+				if !ok || len(metadata.GetDirectlyRelatedUserTypes()) == 0 {
+					return serverErrors.AssignableRelationHasNoTypes(td.GetType(), relation)
+				}
+			} else {
+				if ok && len(metadata.GetDirectlyRelatedUserTypes()) > 0 {
+					return serverErrors.NonassignableRelationHasAType(td.GetType(), relation)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func isAssignable(rewrite *openfgapb.Userset) bool {
+	switch rw := rewrite.GetUserset().(type) {
+	case *openfgapb.Userset_This:
+		return true
+	case *openfgapb.Userset_Union:
```

Comment:
If we used `UsersetRewrite_SetOp` here instead of individual `Union`, `Intersection`, `Difference` this would collapse into

```
case *openfgapb.UsersetRewrite_SetOp:
    for _, child := rw.GetSetOperation().GetChild() {
       if isAssignable(child) { return true }
    }
```
Another +1 opportunity for that refactor at some point 😄